### ISSUE TITLE
Add monitoring to the server

### ISSUE THREAD

**Issue by wilbertom** [sc3/cookcountyjail#67]
Log server queries.

**Comment by eads** [comment]
Do we still need to do this?

**Comment by nwinklareth** [comment]
Yes,at least two:
- webserver
- scrapping program

Failure notifications should be sent to the supreme coding mailing list so
one of us can address the issue.

On Sat, Jul 27, 2013 at 11:14 AM, David Eads notifications@github.comwrote:

> Do we still need to do this?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/sc3/cookcountyjail/issues/67#issuecomment-21667926
> .

## 

Regards

## Norbert

Norbert Winklareth

**Comment by eads** [comment]
Yeah, let's split this out into a couple of tickets. I'm thinking
@bepetersn is probably the person for the job. A "health" API endpoint is
the other thing we should add.

**Comment by bepetersn** [comment]
From what I can see, it's not possible to monitor our web server, or our API if the former is down. 

So, the only way to have a full health API is thus to run it on a different server, which is a sort of industry standard from what I have read.

I can't initiate getting a new web server currently, but it ought to be simple to implement these basic functionalities once we have the server set up. 

It might further make sense, if we had a second server, to have the first one which currently runs the scraper to submit POST requests to the second. For instance, Norbert's scraper could just send records of its activity this way.

A rather far off thought, but worth sharing, I thought.

**Comment by nwinklareth** [comment]
There are number of types of monitoring required:

1) Should be a /dashboard that displays information on how the webserver has performed
2) If the webserver goes down then it needs to be restarted
3) Performance monitoring of webserver, detects poor performance and has it or server machine restarted
4) server machine monitoring, signals someone if server machine is no longer responding

The first two can be done on the server machine, #2 can be done with a tool like supervisor. The last two require external monitoring and that typically involves cost.

**Comment by bepetersn** [comment]
Do we still have an interest in keeping track of server queries, like @wilbertom originally suggested with this issue? I guess the idea would be that we could keep track of what requests people are trying to make on our site?

**Comment by bepetersn** [comment]
Reposting this comment of mine from #332. My conclusion there is that we need to look into using a message queue like [python-rq](http://python-rq.org/) for our server / scraper / DB monitoring needs. I go on to say:
- We've talked for awhile about having some part of the scraper always running on the server. I think this is the answer to the need for that "always-on" component.
- We still need an API that shows the status of our project. When I conceptualized this before, I couldn't think how to keep this up-to-date constantly (as opposed to just after our once-a-day programs run), and thinking that we wouldn't need anything other than this 80-90% of the time, I ignored the issue. But if we have a program constantly running on the server, IT can update the status API.
- Our status API then might just be a shallow representation that gets updated by this Monitor/Controller program.
- We could still keep a history of AdminTasks if desired, which the Monitor could pull from to keep the status API up to date. This might not be an essential component, however, and we could try to do without it during a first iteration.
